File: csharp/affine-cipher/AffineCipher.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using Mod = System.Func<int, int, int, int>;

public static class AffineCipher
{
    private static int m = 26;

    public static string Encode(string plainText, int a, int b)
    {
        var inverse = FindInverse(a, m);

        return plainText
            .Where(char.IsLetterOrDigit)
            .Select((c) => Encrypt(c, a, b))
            .Select((e, i) => (i > 0 && i % 5 == 0) ? $" {e}" : $"{e}")
            .Aggregate(
                string.Empty,
                (accumulator, next) => $"{accumulator}{next}"
            );
    }

    public static string Decode(string cipheredText, int a, int b)
    {
        var inverse = FindInverse(a, m);

        return cipheredText
            .Where(char.IsLetterOrDigit)
            .Select((c) => Decrypt(c, inverse, b))
            .Aggregate(
                string.Empty,
                (accumulator, next) => $"{accumulator}{next}"
            );
    }

    private static char Decrypt(char c, int a, int b) =>
        Cipher(c, a, b, GetDecryptMod);

    private static char Encrypt(char c, int a, int b) =>
        Cipher(c, a, b, GetEncryptMod);

    private static char Cipher(char c, int a, int b, Mod GetMod)
    {
        if (char.IsDigit(c))
            return c;

        if (char.IsUpper(c))
            c = char.ToLower(c);

        var x = c - 97;
        var mod = GetMod(a, b, x);

        if (mod < 0)
            mod = mod + m;

        return (char)(mod + 97);
    }

    private static Mod GetEncryptMod =>
        (a, b, x) => (a * x + b) % m;

    private static Mod GetDecryptMod =>
        (a, b, x) => ((x - b) * a) % m;

    private static int FindInverse(int a, int b)
    {
        var (x0, x1, y0, y1) = (1, 0, 0, 1);

        while (a != 0)
        {
            var q = b / a;
            (b, a) = (a, b % a);
            (x0, x1) = (x1, x0 - q * x1);
            (y0, y1) = (y1, y0 - q * y1);
        }

        if (b > 1)
            throw new ArgumentException("a and m must be coprime.");

        return y0;
    }
}

File: fsharp/affine-cipher/AffineCipher.fs
module AffineCipher

let m = 26

let findInverse u v =
    let rec findInverseImpl a b x0 x1 y0 y1 =
        if a = 0 && b > 1 then
            invalidArg "a" "a must be coprime with 26"
        elif a = 0 then
            y0
        else
            let q = b / a
            let bnew, anew = a, b % a
            let x0new, x1new = x1, x0 - q * x1
            let y0new, y1new = y1, y0 - q * y1

            findInverseImpl anew bnew x0new x1new y0new y1new

    findInverseImpl u v 1 0 0 1

let moduloDecrypt a b x =
    ((x - b) * a) % m

let moduloEncrypt a b x =
    (a * x + b) % m

let rec private compute f a b c =
    if not (System.Char.IsLetter(c)) then
        c
    elif System.Char.IsUpper(c) then
        compute f a b (System.Char.ToLower(c))
    else
        let x = (int) c - 97
        let modulo = f a b x

        let result =
            if modulo < 0 then (modulo + m) else modulo

        (char) (result + 97)

let rec private addSpaces (source: char list) =
    match source with
    | e1 :: e2 :: e3 :: e4 :: e5 :: tail when (List.length tail > 0) ->
        e1
        :: e2
        :: e3
        :: e4
        :: e5
        :: ' '
        :: (addSpaces tail)
    | _ -> source

let decode a b cipheredText =
    let inverse = findInverse a m
    cipheredText
    |> Seq.filter System.Char.IsLetterOrDigit
    |> Seq.map (compute moduloDecrypt inverse b)
    |> Seq.toArray
    |> System.String

let encode a b plainText =
    let inverse = findInverse a m

    plainText
    |> Seq.toList
    |> List.filter System.Char.IsLetterOrDigit
    |> List.map (compute moduloEncrypt a b)
    |> addSpaces
    |> List.toArray
    |> System.String
